File: 2872-maximum-number-of-k-divisible-components/2872-maximum-number-of-k-divisible-components.java
class Solution {
    public int maxKDivisibleComponents (int n, int [] [] edges, int [] values, int k) {
        List <List <Integer>> adj = new ArrayList <> ();
        for (int i = 0; i < n; i++) adj.add (new ArrayList <> ());
        for (int edge [] : edges) {
            adj.get (edge [1]).add (edge [0]);
            adj.get (edge [0]).add (edge [1]);
        } 
        int res [] = new int [1];
        dfs (adj, values, 0, -1, k, res);
        return res [0];
    }
    
    private int dfs (List <List <Integer>> adj, int values [], int node, int parent, int k, int res []) {
        int sum = values [node];
        for (int temp : adj.get (node)) {
            if (temp != parent) 
                sum += dfs (adj, values, temp, node, k, res);
        }
        if (sum % k == 0) res [0] ++;
        return sum % k;
    }
}

File: 2872-maximum-number-of-k-divisible-components/2872-maximum-number-of-k-divisible-components.py
class Solution:
    def maxKDivisibleComponents (self, n : int, edges : List [List [int]], values : List [int], k : int) -> int:
        adj = [[] for _ in range (n)]
        for a, b in edges:
            adj [a].append (b)
            adj [b].append (a)
        self.res = 0
        def dfs (node, parent):
            s = values [node]
            for nextt in adj [node]:
                if nextt != parent:
                    s += dfs (nextt, node)
            if s % k == 0:
                self.res += 1
            return s % k
        dfs (0, -1)
        return self.res